guides/release/using-typescript/ts-and-ember.md

```diff
@@ -0,0 +1,317 @@
+   get(get(get(someObject, 'deeply'), 'nested'), 'key');
+   ```
+
+2. **Using `// @ts-ignore`.** This will _not do any type-checking_, but is useful for the cases where you are intentionally checking a path which may be `null` or `undefined` anywhere long it.
+
+   ```typescript
+   // @ts-ignore
+   get(someObject, 'deeply.nested.key');
+   ```
+
+   It's usually best to include an explanation of _why_ you're ignoring a lookup!
+
+### Service and controller injections
+
+Ember does service and controller lookups with the `inject` functions at runtime, using the name of the service or controller being injected up as the default value—a clever bit of metaprogramming that makes for a nice developer experience. TypeScript cannot do this, because the name of the service or controller to inject isn't available at compile time in the same way.
+
+The officially supported method for injections with TypeScript uses _decorators_.
+
+```typescript
+// my-app/services/my-session.ts
+import Service from '@ember/service';
+import RSVP from 'rsvp';
+
+export default class MySession extends Service {
+  login(email: string, password: string): RSVP.Promise<string> {
+    // login and return the confirmation message
+  }
+}
+
+declare module '@ember/service' {
+  interface Registry {
+    'my-session': MySession;
+  }
+}
+```
+
+Then we can use the service as we usually would with a decorator, but adding a type annotation to it so TypeScript knows what it's looking at:
+
+```typescript
+// my-app/components/user-profile.ts
+import Component from '@ember/component';
+import { inject as service } from '@ember/service';
+
+import MySession from 'my-app/services/my-session';
+
+export default class UserProfile extends Component {
+  @service declare mySession: MySession;
+
+  login(email: string, password: string) {
+    this.mySession.login(email, password);
+  }
+}
+```
+
+Note that we need the `MySession` type annotation this way, but we _don't_ need the string lookup (unless we're giving the service a different name than the usual on the class, as in Ember injections in general). Without the type annotation, the type of `session` would just be `any`. This is because decorators are not allowed to modify the types of whatever they decorate. As a result, we wouldn't get any type-checking on that `session.login` call, and we wouldn't get any auto-completion either. Which would be really sad and take away a lot of the reason we're using TypeScript in the first place!
+
+Also notice [the `declare` property modifier](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier). This tells TypeScript that the property will be configured by something outside the class (in this case, the decorator), and guarantees it emits spec-compliant JavaScript.
+
+(This also holds true for all other service injections, computed property macros, and Ember Data model attributes and relationships.)
+
```

Comment:
~~~suggestion

Finally, you may have noticed the `declare module` at the bottom of the example `MySession` definition:

```typescript
declare module '@ember/service' {
  interface Registry {
    'my-session': MySession;
  }
}
```

This definition allows for type-safe lookups with other Ember dependency injection APIs. For example, [the `Owner.lookup` method](https://api.emberjs.com/ember/5.2/classes/Owner#2-method) uses this "registration"—a mapping from the string `'my-session'` to the service type, `MySession`—to provide the correct type:

```typescript
function dynamicLookup(owner: Owner) {
  let mySession = owner.lookup('service:my-session');
  mySession.login("tom@example.com", "password123");
}
```

~~~